### PR TITLE
generate snapshot in background

### PR DESCRIPTION
**Pull request by arcadiushmcts**

### CODE REVIEW

File: aggregate-snapshot/aggregate-snapshot-service/src/main/java/uk/gov/justice/services/eventsourcing/source/core/snapshot/DefaultSnapshotService.java [hunk 1]
```diff
@@ -55,6 +56,21 @@ public <T extends Aggregate> void attemptAggregateStore(final UUID streamId, fin
...
+
+    public <T extends Aggregate> boolean storeAggregateSimply(final UUID streamId, final long streamVersionId, final T aggregate) {
+        try {
+            logger.debug("Storing snapshot of aggregate: {}, streamId: {}, version: {}", aggregate.getClass().getSimpleName(), streamId, streamVersionId);
+            final AggregateSnapshot<T> aggregateSnapshot = new AggregateSnapshot<>(streamId, streamVersionId, aggregate, null);//createdAt is added before saving
+            final boolean storedSuccessfully = snapshotRepository.storeSnapshot(aggregateSnapshot);
+            logger.debug("Stored successfully {}", storedSuccessfully);
+            return storedSuccessfully;
+        } catch (SerializationException e) {
+            logger.error("Error creating snapshot for %s".formatted(streamId), e);
```
Comment (allanmckenzie) [inline aggregate-snapshot/aggregate-snapshot-service/src/main/java/uk/gov/justice/services/eventsourcing/source/core/snapshot/DefaultSnapshotService.java:68]: Didn't know you could do this kind of formatting. Nice :-)

File: aggregate-snapshot/aggregate-snapshot-domain/src/test/java/uk/gov/justice/domain/snapshot/AggregateSnapshotTest.java [hunk 2]
```diff
@@ -26,7 +26,7 @@ public class AggregateSnapshotTest {
     public void shouldCreateAnAggregateSnapshot() throws Exception {
         final TestAggregate aggregate = new TestAggregate("STATE1");
 
-        final AggregateSnapshot<TestAggregate> snapshot = new AggregateSnapshot<>(STREAM_ID, VERSION_ID, aggregate);
+        final AggregateSnapshot<TestAggregate> snapshot = new AggregateSnapshot<>(STREAM_ID, VERSION_ID, aggregate, null);
```
Comment (allanmckenzie) [inline aggregate-snapshot/aggregate-snapshot-domain/src/test/java/uk/gov/justice/domain/snapshot/AggregateSnapshotTest.java:29]: I'm not sure I like passing in null like this. I've found nulls usually lead to tears and pain. Can this be an Optional instead?

File: aggregate-snapshot/aggregate-snapshot-domain/src/main/java/uk/gov/justice/domain/snapshot/AggregateSnapshot.java [hunk 3]
```diff
@@ -18,25 +19,29 @@
...
     }
 
-    public AggregateSnapshot(final UUID streamId, final Long versionId, final Class<T> type, final byte[] aggregateByteRepresentation) {
-        this(streamId, versionId, type.getName(), aggregateByteRepresentation);
+    public AggregateSnapshot(final UUID streamId, final Long versionId, final Class<T> type, final byte[] aggregateByteRepresentation, final ZonedDateTime createdAt) {
+        this(streamId, versionId, type.getName(), aggregateByteRepresentation, createdAt);
     }
 
-    public AggregateSnapshot(final UUID streamId, final Long versionId, final String type, final byte[] aggregateByteRepresentation) {
+    public AggregateSnapshot(final UUID streamId, final Long versionId, final String type, final byte[] aggregateByteRepresentation, final ZonedDateTime createdAt) {
```
Comment (allanmckenzie) [inline aggregate-snapshot/aggregate-snapshot-domain/src/main/java/uk/gov/justice/domain/snapshot/AggregateSnapshot.java:39]: If the aggregateByteRepresentation byte array and createdAt date are linked, should they be but into a new single Object, AggregateByteArray or something. Also, if it's not always been created, should that Object be Optional?

File: aggregate-snapshot/aggregate-snapshot-service/src/main/java/uk/gov/justice/services/core/aggregate/DefaultAggregateService.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    @Inject
+    JsonObjectToObjectConverter jsonObjectToObjectConverter;
+
+    @Inject
+    private AsyncSnapshotService asyncSnapshotService;
+
+    @Inject
+    @Value(key = "snapshot.background.saving.threshold", defaultValue = "" + SNAPSHOT_BACKGROUND_SAVING_THRESHOLD)
+    private long snapshotBackgroundSavingThreshold;
```
[the remark this answers is not in the document]
Comment (arcadius) [reply aggregate-snapshot/aggregate-snapshot-service/src/main/java/uk/gov/justice/services/core/aggregate/DefaultAggregateService.java:47]: Thanks @santhosh-kumar3   for the comment.
On SIT:

```          <simple
                name="java:/app/material-service/material.sasUrlExpiryInMinutes"
                type="java.lang.String"
                value="30"
               />
```

and the java code looks like this:

``` 
@Inject
@Value(key = "material.sasUrlExpiryInMinutes", defaultValue = "30")
 private long expiryMinutes;
```

Also just for testing this, I have edited the docker project to add JNDI value 
```
<simple name="java:/app/results-service/snapshot.background.saving.threshold" value="10000" type="java.lang.String"/>
```

and I have changed the code to log the value 

```
logger.info("==== snapshotBackgroundSavingThreshold is {}", snapshotBackgroundSavingThreshold);
```

after redeploying results, I can see the output

```
==== snapshotBackgroundSavingThreshold is 10000
```

so, the automatic long/in conversion seems to be working

File: aggregate-snapshot/aggregate-snapshot-domain/src/main/java/uk/gov/justice/domain/snapshot/AggregateSnapshot.java (hunk 3 above)
Comment (arcadius) [reply aggregate-snapshot/aggregate-snapshot-domain/src/main/java/uk/gov/justice/domain/snapshot/AggregateSnapshot.java]: fixed

File: aggregate-snapshot/aggregate-snapshot-domain/src/test/java/uk/gov/justice/domain/snapshot/AggregateSnapshotTest.java (hunk 2 above)
Comment (arcadius) [reply aggregate-snapshot/aggregate-snapshot-domain/src/test/java/uk/gov/justice/domain/snapshot/AggregateSnapshotTest.java]: fixed

**Pull request merged**